### PR TITLE
[PyTorch] Fix a bunch of structured kernel refcounting

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71140

Structured kernels need to use the borrowing variants of the build APIs to TensorIterator. (I am working on a debug check for this, but it is currently too strict, and relaxing it does not catch these bugs.)

Differential Revision: [D33520003](https://our.internmc.facebook.com/intern/diff/D33520003/)

### CODE REVIEW

File: aten/src/ATen/native/Pow.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ TORCH_META_FUNC2(pow, Tensor_Scalar) (const Tensor& base, const Scalar& exp) {
               "Integers to negative integer powers are not allowed.");
 
   auto common_dtype = at::result_type(base, exp);
-  build_unary_op(maybe_get_output(), base.to(common_dtype));
+  build_output_borrowing_argument_owning_unary_op(maybe_get_output(), base.to(common_dtype));
```
Comment (bdhirsh) [inline aten/src/ATen/native/Pow.cpp:23]: oof okay. So is the idea that:
- in general for structured kernels, we always borrow the input and output tensors

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ void mkldnn_matmul(
 bool use_mkldnn_bf16_matmul(
     const Tensor& mat1,
     const Tensor& mat2,
-    const c10::optional<Tensor>& result_opt){
+    const Tensor& result_opt){
```
Comment (bdhirsh) [inline aten/src/ATen/native/mkldnn/Matmul.cpp:22]: why are we removing these optionals?

**Review by bdhirsh — APPROVED** [review]
Nice catch! This probably came up before, but quick question:

In this case, usage of the owning TI constructor caused the bug, where we should have used the borrowing TI constructor. It it always the case that the borrowing construct is "safer" than the owning constructor. And if so, do you think we should make the default "borrowing" instead of "owning"? I have a feeling this came up before and the answer was that we shouldn't since that would change the behavior of all the existing TI ops.

**Comment by ngimel** [comment]
It's interesting that internally a lot of tests are legitimately failing with 
```
==11581==ERROR: AddressSanitizer: heap-use-after-free on address 0x609000002d40 at pc 0x7ff4a6099ea2 bp 0x7fff62a6c740 sp 0x7fff62a6c738
```
and OSS CI is fine. cc @dagitses for what tests should be added to catch this type of failure.

**Pull request closed without merging**